data_pipeline/google_analytics/ga_pipeline.py

```diff
@@ -0,0 +1,180 @@
+                        metric_headers, values.get('values')
+                ):
+                    bq_json_formatted_record[
+                        standardize_field_name(metric_header.get('name'))
+                    ] = value
+
+            yield bq_json_formatted_record
+
+
+def etl_google_analytics(
+        ga_config: GoogleAnalyticsConfig,
+        start_date: datetime,
+        end_date: datetime = None,
+):
+    current_timestamp_as_string = get_current_timestamp_as_string()
+    analytics = GoogleAnalyticsClient()
+    end_date = end_date or datetime.now()
+    from_date = start_date.strftime("%Y-%m-%d")
+    to_date = end_date.strftime("%Y-%m-%d")
+    dimensions = [
+        {
+            'name': dim_name
+        } for dim_name in ga_config.dimensions
+    ]
+    metrics = [
+        {
+            'expression': metric_name
+        } for metric_name in ga_config.metrics
+    ]
+    page_token = None
+    while True:
+
+        response = analytics.get_report(
+            date_ranges=[{'startDate': from_date, 'endDate': to_date}],
+            view_id=ga_config.ga_view_id,
+            metrics=metrics,
+            dimensions=dimensions,
+            page_token=page_token
+        )
+        reports = response.get('reports', [])
+        # api can return multiple reports, however,
+        # this implementation considers single report
+        page_token = (
+            reports[0].get('nextPageToken')
+            if len(reports) == 1 else None
+        )
+
+        transformed_response_with_provenance = (
+            add_provenance(
+                transform_response_to_bq_compatible_record(
+                    response
+                ),
+                ga_config.import_timestamp_field_name,
+                current_timestamp_as_string,
+                ga_config.record_annotations
+            )
+        )
+        with NamedTemporaryFile() as temp_file:
+            temp_file_name = temp_file.name
+
+            write_result_to_file(
+                json_list=transformed_response_with_provenance,
+                full_temp_file_location=temp_file_name,
+                write_mode="w"
+            )
+            load_written_data_to_bq(
+                ga_config=ga_config,
+                file_location=temp_file_name
+            )
+        update_state(
+            start_date,
+            ga_config.state_s3_bucket_name,
+            ga_config.state_s3_object_name
+        )
+        if not page_token:
+            break
+
+
+def add_provenance(
+        ga_records: Iterable[dict],
+        timestamp_field_name: str,
+        current_etl_time: str,
+        record_annotation: dict
+):
+    for record in ga_records:
+        provenance = {
+            'provenance': {
+                timestamp_field_name: current_etl_time,
+                'annotation': record_annotation
+            }
+        }
+
+        yield {
+            **record,
+            **provenance
+        }
+
+
+def write_result_to_file(
```

Comment:
the google sheets pipeline seem to have a similar method. Might be worth moving to a common module. I guess the name could reflect that we are writing jsonl.